File: ruby/dependency_tracker.rb
require 'find'

def scan_for_dependencies(directory = '.')
  dependencies = []

  Find.find(directory) do |path|
    if File.extname(path) == '.rb'
      File.readlines(path).each do |line|
        match = line.match(/^\s*require\s+['"](.+?)['"]/)
        dependencies << match[1] if match && !match[1].start_with?('.')
      end
    end
  end

  dependencies.uniq
end

def generate_gemfile(dependencies, output_path = 'Gemfile')
  gemfile_content = <<~HEREDOC
    source "https://rubygems.org"

    gemspec
  HEREDOC

  dependencies.each do |dep|
    gemfile_content << "gem '#{dep}'\n"
  end

  File.write(output_path, gemfile_content)
  puts "Gemfile generated successfully at #{output_path}."
end

if __FILE__ == $0
  directory = ARGV.find { |arg| arg.start_with?('--dir=') }&.split('=')&.last || '.'
  output_path = ARGV.find { |arg| arg.start_with?('--output=') }&.split('=')&.last || 'Gemfile'

  dependencies = scan_for_dependencies(directory)
  generate_gemfile(dependencies, output_path)
end

File: autodep/dependency_tracker.py
import os
import argparse

def scan_for_dependencies(directory='.'):
    dependencies = set()
    for root, _, files in os.walk(directory):
        for file in files:
            if file.endswith('.py'):
                with open(os.path.join(root, file)) as f:
                    for line in f:
                        if line.startswith('import') or line.startswith('from'):
                            dependencies.add(line.strip().split()[1])
    return dependencies

def generate_requirements_file(dependencies, output_path='requirements.txt'):
    with open(output_path, 'w') as f:
        for dep in dependencies:
            f.write(f"{dep}\n")
    print(f"Requirements file generated successfully at {output_path}.")

def main():
    parser = argparse.ArgumentParser(description="AutoDep - Automatically scan your Python project for dependencies and generate a requirements.txt file.")
    parser.add_argument('--dir', type=str, default='.', help='Directory to scan for dependencies (default: current directory).')
    parser.add_argument('--output', type=str, default='requirements.txt', help='Path to save the generated requirements.txt file (default: requirements.txt).')
    args = parser.parse_args()

    dependencies = scan_for_dependencies(args.dir)
    generate_requirements_file(dependencies, args.output)

if __name__ == "__main__":
    main()
